### ISSUE TITLE
Option for mouse sensitivity types

### ISSUE THREAD

**Issue by CapnClever** [fx02/cndoom#12]
I've heard from a few runners that the decision to choose (and stick to) a particular source port can be significantly affected by which mouse sensitivity they become comfortable with. Practically speaking, this is often the difference between "vanilla" and "prboom+": although cndoom has the option to individually set horizontal and vertical speed/threshold/acceleration, it is still calculated similar to the original game, whereas other ports have diverged.

In short, I'm suggesting a feature to allow users to choose between mouse sensitivity options. An initial implementation would switch merely between cndoom's current option (maybe "Default") and "PrBoom+", which would use that source's mouse handling. (In the future, this could be expanded to include, for example, ZDoom's handling, with its unique allowance to adjust mouse strafe speed separate from turn speed.) All of the input-relevant code should be localized in `i_video` (or `i_input` in the sdl2-branch when that becomes relevant), and the rest of the work would be setup/config changes to allow for the switch in-menu (and relevant settings for that sensitivity type altered).

While this doesn't adhere to a vanilla-like standard, I believe it's more important in the context of a speedrun-friendly source port for users to be comfortable enough to play at their best. (From a use-case perspective, this makes cndoom more accessible to the existing community.) Additionally, as this is all occurring before the ticcmd is constructed, this kind of option should never cause demo sync problems, so that's nothing to worry about.

**Comment by fragglet** [comment]
What differences are there in terms of logic between the different sensitivity options?

**Comment by CapnClever** [comment]
To start, looks like I was very wrong about the data being localized: adjustments occur in up to three separate functions depending on context. None of these are after `G_BuildTicCmd()`, though, so I still don't think changes would cause desyncs.

**Chocolate Doom**

Event read occurs in `i_video/I_ReadMouse()`. This function calls `SDL_GetRelativeMouseSpeed()` to get initial values, then calls `AccelerateMouse(speed)` on each axis (x stored as `ev.data2` is and y as `ev.data3`), basic formula shown here:
```
ev.data2 = (int)((x - mouse_threshold) * mouse_acceleration + mouse_threshold);
ev.data3 = -(int)((y - mouse_threshold) * mouse_acceleration + mouse_threshold);
```
From there it's changed in near the end of `g_game/G_Responder()`, where a non-zero sensitivity value ("speed" in setup) allows this to occur (paraphrased):
```
mousex = ev->data2*(mouseSensitivity+5)/10;
mousey = ev->data3*(mouseSensitivity+5)/10;
```
where "mouse" will be the value that the tic command starts with. Finally, `G_BuildTiccmd()` does some last-second alterations when applying the values to actual motion:
```
forward += mousey;
if (strafe) 
  side += mousex*2; 
else 
  cmd->angleturn -= mousex*0x8; 
```

Only difference in CNDoom is that x/y values have their own sensitivity, threshold, and acceleration values.

**PrBoom+**

Event read occurs in `i_video/I_GetEvent()` as part of the switch to check event type:
```
case SDL_MOUSEMOTION:
...
  event.data2 = Event->motion.xrel << 4;
  event.data3 = -Event->motion.yrel << 4;
...
```
where `Event` is the raw SDL_Event object from a poll. In `g_game/G_Responder()` the formula goes like this:
```
mousex += (AccelerateMouse(ev->data2)*(mouseSensitivity_horiz))/10;
mousey += (AccelerateMouse(ev->data3)*(mouseSensitivity_vert))/40;

// where AccelerateMouse(val) is equivalent to:
M_DoubleToInt(pow((double)val, (double)((mouse_acceleration/100.f) + 1.f)))
```
In `g_game/G_BuildTiccmd()`, the alterations to motion are as follows:
```
forward += mousey;
if (strafe)
  side += mousex / 4;
else
  cmd->angleturn -= mousex;
```

In summary, the details are different all over the place: sometimes by a constant multiple, other times quite drastically. Should be possible to incorporate both, though.

**Comment by fx02** [comment]
I have nothing against other mouse sensitivity types, in DOS era we used few tools to adjust mouse to our needs so more sensitivity manipulation is a yes. In terms what community needs is quickstart (adjust mouse angle before game started) that existed in Vanilla and that still does not exist in Chocolate.

**Comment by CapnClever** [comment]
> I have nothing against other mouse sensitivity types, in DOS era we used few tools to adjust mouse to our needs so more sensitivity manipulation is a yes.

Awesome, I'll get started looking into an implementation.

> In terms what community needs is quickstart (adjust mouse angle before game started) that existed in Vanilla and that still does not exist in Chocolate.

This looks like an inconsistent behavior WRT vanilla altogether: you can adjust the angle before the screen wipe finishes, but in Chocolate there appears to be a limit as to how far it goes. I'll write it up as an issue upstream.

**Comment by fx02** [comment]
There is already issue opened and we had it solved in first cndoom release but was not so pretty :)
